radix-engine-interface/src/data/value_serializer.rs

```diff
@@ -0,0 +1,667 @@
+    };
+
+    #[scrypto(TypeId, Encode, Decode)]
+    pub struct Sample {
+        pub a: ResourceAddress,
+    }
+
+    pub fn assert_json_eq<T: Serialize>(actual: T, expected: Value) {
+        let actual = to_value(&actual).unwrap();
+        if actual != expected {
+            panic!(
+                "Mismatching JSON:\nActual:\n{}\nExpected:\n{}\n",
+                to_string(&actual).unwrap(),
+                to_string(&expected).unwrap()
+            );
+        }
+    }
+
+    #[test]
+    #[cfg(feature = "serde")] // Workaround for VS Code "Run Test" feature
+    fn test_address_encoding_no_network() {
+        let value = RADIX_TOKEN;
+
+        let no_network_radix_token_address = RADIX_TOKEN.display(NO_NETWORK).to_string();
+
+        let expected = json!(no_network_radix_token_address);
+        let expected_invertible =
+            json!({ "type": "ResourceAddress", "value": no_network_radix_token_address });
+
+        assert_simple_json_matches(&value, ValueFormattingContext::no_context(), expected);
+        assert_invertible_json_matches(
+            &value,
+            ValueFormattingContext::no_context(),
+            expected_invertible,
+        );
+    }
+
+    #[test]
+    #[cfg(feature = "serde")] // Workaround for VS Code "Run Test" feature
+    fn test_address_encoding_with_network() {
+        let value = RADIX_TOKEN;
+        let encoder = Bech32Encoder::for_simulator();
+
+        let radix_token_address = RADIX_TOKEN.display(&encoder).to_string();
+
+        let expected_simple = json!(radix_token_address);
+        let expected_invertible =
+            json!({ "type": "ResourceAddress", "value": radix_token_address });
+
+        assert_simple_json_matches(&value, &encoder, expected_simple);
+        assert_invertible_json_matches(&value, &encoder, expected_invertible);
+    }
+
+    #[test]
+    #[cfg(feature = "serde")] // Workaround for VS Code "Run Test" feature
+    fn test_complex_encoding_with_network() {
+        use crate::{
+            core::Expression,
+            math::{Decimal, PreciseDecimal},
+        };
+
+        let encoder = Bech32Encoder::for_simulator();
+        let radix_token_address = RADIX_TOKEN.display(&encoder).to_string();
+
+        let value = ScryptoValue::Tuple {
+            fields: vec![
+                SborValue::U8 { value: 5 },
+                SborValue::I32 { value: -5 },
+                SborValue::U64 { value: u64::MAX },
+                SborValue::U128 {
+                    value: 9912313323213,
+                },
+                SborValue::Array {
+                    element_type_id: SborTypeId::U8,
+                    elements: vec![SborValue::U8 { value: 0x3a }, SborValue::U8 { value: 0x92 }],
+                },
+                SborValue::Array {
+                    element_type_id: SborTypeId::U32,
+                    elements: vec![SborValue::U32 { value: 153 }, SborValue::U32 { value: 62 }],
+                },
+                SborValue::Tuple {
```

Comment:
Looks like we're missing enum test here. Would also like one instance for each custom type, so we're less likely to break it.